mls-rs-uniffi/src/lib.rs

```diff
@@ -0,0 +1,524 @@
+
+    async fn index_to_identity(
+        &self,
+        index: u32,
+    ) -> Result<mls_rs_core::identity::SigningIdentity, Error> {
+        let group = self.inner().await;
+        let member = group
+            .member_at_index(index)
+            .ok_or(MlsError::InvalidNodeIndex(index))?;
+        Ok(member.signing_identity)
+    }
+}
+
+/// Extract the basic credential identifier from a  from a key package.
+fn signing_identity_to_identifier(
+    signing_identity: &mls_rs_core::identity::SigningIdentity,
+) -> Result<Vec<u8>, Error> {
+    match &signing_identity.credential {
+        identity::Credential::Basic(credential) => Ok(credential.identifier.clone()),
+        _ => Err(MlsError::RequiredCredentialNotFound(BasicCredential::credential_type()).into()),
+    }
+}
+
+#[uniffi::export]
+impl Group {
+    /// Perform a commit of received proposals (or an empty commit).
+    ///
+    /// TODO: ensure `path_required` is always set in
+    /// [`MlsRules::commit_options`](`mls_rs::MlsRules::commit_options`).
+    ///
+    /// Returns the resulting commit message. See
+    /// [`mls_rs::Group::commit`] for details.
+    pub async fn commit(&self) -> Result<Message, Error> {
+        let mut group = self.inner().await;
+        let commit_output = group.commit(Vec::new()).await?;
+        Ok(commit_output.commit_message.into())
+    }
+
+    /// Commit the addition of a member.
+    ///
+    /// The member is representated by the key package in `member`.
+    /// The result is the welcome message to send to this member.
+    ///
+    /// See [`mls_rs::group::CommitBuilder::add_member`] for details.
+    pub async fn add_member(&self, member: Arc<Message>) -> Result<CommitOutput, Error> {
+        let member = arc_unwrap_or_clone(member);
+        let mut group = self.inner().await;
+        let commit_output = group
+            .commit_builder()
+            .add_member(member.inner)?
+            .build()
+            .await?;
+        Ok(commit_output.into())
+    }
+
+    /// Propose to add a member to this group.
+    ///
+    /// The member is representated by the key package in `member`.
+    /// The result is the welcome message to send to this member.
+    ///
+    /// See [`mls_rs::Group::propose_add`] for details.
+    pub async fn propose_add_member(&self, key_package: Arc<Message>) -> Result<Message, Error> {
+        let key_package = arc_unwrap_or_clone(key_package);
+        let mut group = self.inner().await;
+        let mls_message = group.propose_add(key_package.inner, Vec::new()).await?;
+        Ok(mls_message.into())
+    }
+
+    /// Propose and commit the removal of a member.
+    ///
+    /// The member is representated by the key package in `member`.
+    ///
+    /// See [`mls_rs::group::CommitBuilder::remove_member`] for details.
+    pub async fn remove_member(&self, member: Arc<SigningIdentity>) -> Result<CommitOutput, Error> {
+        let identifier = signing_identity_to_identifier(&member.inner)?;
```

Comment:
Can it be `member: &SigningIdentity`?

Reply:
Yes, much better!

If/when `Group::propose_add` and `ClientBuilder::add_member` would stop taking ownership of the key package passed to them, then we can also simplify the signatures for `add_member` and `propose_add_member`.